pkg/volume/storageos/storageos.go

```diff
@@ -0,0 +1,705 @@
+
+func (b *storageosUnmounter) GetPath() string {
+	return getPath(b.podUID, b.volNamespace, b.volName, b.pvName, b.plugin.host)
+}
+
+// Unmounts the bind mount, and detaches the disk only if the PD
+// resource was the last reference to that disk on the kubelet.
+func (b *storageosUnmounter) TearDown() error {
+	if len(b.volNamespace) == 0 || len(b.volName) == 0 {
+		glog.Warningf("volNamespace: %q, volName: %q not set, skipping TearDown", b.volNamespace, b.volName)
+		return fmt.Errorf("pvName not specified for TearDown, waiting for next sync loop")
+	}
+	// Unmount from pod
+	mountPath := b.GetPath()
+
+	err := b.TearDownAt(mountPath)
+	if err != nil {
+		glog.Errorf("Unmount from pod failed: %v", err)
+		return err
+	}
+
+	// Find device name from global mount
+	globalPDPath := makeGlobalPDName(b.plugin.host, b.pvName, b.volNamespace, b.volName)
+	devicePath, _, err := mount.GetDeviceNameFromMount(b.mounter, globalPDPath)
+	if err != nil {
+		glog.Errorf("Detach failed when getting device from global mount: %v", err)
+		return err
+	}
+
+	// Unmount from plugin's disk global mount dir.
+	err = b.TearDownAt(globalPDPath)
+	if err != nil {
+		glog.Errorf("Detach failed during unmount: %v", err)
+		return err
+	}
+
+	// Detach loop device
+	err = b.manager.DetachVolume(b, devicePath)
+	if err != nil {
+		glog.Errorf("Detach device %s failed for volume %s: %v", devicePath, b.pvName, err)
+		return err
+	}
+
+	glog.V(4).Infof("Successfully unmounted StorageOS volume %s and detached devices", b.pvName)
+
+	return nil
+}
+
+// Unmounts the bind mount, and detaches the disk only if the PD
+// resource was the last reference to that disk on the kubelet.
+func (b *storageosUnmounter) TearDownAt(dir string) error {
+	if err := util.UnmountPath(dir, b.mounter); err != nil {
+		glog.V(4).Infof("Unmounted StorageOS volume %s failed with: %v", b.pvName, err)
+	}
+	if err := b.manager.UnmountVolume(b); err != nil {
+		glog.V(4).Infof("Mount reference for volume %s could not be removed from StorageOS: %v", b.pvName, err)
+	}
+	return nil
+}
+
+type storageosDeleter struct {
+	*storageosMounter
+	pvUID types.UID
+}
+
+var _ volume.Deleter = &storageosDeleter{}
+
+func (d *storageosDeleter) GetPath() string {
+	return getPath(d.podUID, d.volNamespace, d.volName, d.pvName, d.plugin.host)
+}
+
+func (d *storageosDeleter) Delete() error {
+	return d.manager.DeleteVolume(d)
+}
+
+type storageosProvisioner struct {
+	*storageosMounter
+	options volume.VolumeOptions
+}
+
+var _ volume.Provisioner = &storageosProvisioner{}
+
+func (c *storageosProvisioner) Provision() (*v1.PersistentVolume, error) {
+
+	adminSecretName := ""
+	adminSecretNamespace := "default"
+
+	// Apply ProvisionerParameters (case-insensitive). We leave validation of
+	// the values to the cloud provider.
+	for k, v := range c.options.Parameters {
+		switch strings.ToLower(k) {
+		case "adminsecretname":
+			adminSecretName = v
+		case "adminsecretnamespace":
```

Comment:
> This is specified in the storage class, right? That's a privileged object to write, and other plugins let you fully specify the coordinates of a secret in storage class parameters
> 
> If you require a secret for a dynamically provisioned PV to be in the namespace of the PVC, you either have to rely on the user providing it (unlikely), or you have to make the provisioner place a secret in the namespace (not good)

@liggitt Good point. I would be fine leaving it as is then, if you're ok with it.

> Would it work if I add a check to enforce it matches either PVC namespace or default?

@croomes At that point might as well remove it as a parameter. So either keep it as is (if @liggitt is ok with that), or remove the `adminsecretnamespace` parameter altogether and use the PVC's namespace. I'm ok with either.